docs/versioning.md

```diff
@@ -0,0 +1,567 @@
+# Versioning
+
+This document describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+## AWS S3 Bucket Versioning
+
+See AWS documentation on Versioning for a description of the Bucket Versioning
+feature:
+
+http://docs.aws.amazon.com/AmazonS3/latest/dev/Versioning.html
+http://docs.aws.amazon.com/AmazonS3/latest/dev/ObjectVersioning.html
+
+## Implementation of Bucket Versioning in S3
+
+### Overview of Metadata and API Component Roles
+
+Each version of an object is stored as a separate key in metadata. The S3 API
+interacts with the metadata backend to store, retrieve, and delete version
+metadata.
+
+The implementation of versioning within the metadata backend is naive. The
+metadata backend does not evaluate any information about bucket or version state
+(whether versioning is enabled or suspended, and whether a version is a null
```

Comment:
Initially the document described most of the feature, including null versions, in the beginning and throughout the document. But we decided it was a lot of irrelevant information to filter through for those who know about the versioning feature and just want to learn how it is implemented. So we removed that information and instead added the links to AWS's description of the feature.